homeassistant/helpers/script.py

```diff
@@ -390,47 +423,108 @@ def async_script_wait(entity_id, from_s, to_s):
+            service,
+            service_data,
+            blocking=True,
+            context=self._context,
+            limit=limit,
+        )
 
-class _BackgroundScriptRun(_ScriptRun):
-    """Manage background Script sequence run."""
+        if limit is None:
+            # No call limit (i.e., potentially starting a fully blocking script call) so
+            # watch for a stop request and, if received, propagate it to the called
+            # scripts by turning them off.
+            task = self._hass.async_create_task(coro)
+            done, pending = await asyncio.wait(
+                {self._stop.wait(), task}, return_when=asyncio.FIRST_COMPLETED,
+            )
+            for pending_task in pending:
+                if pending_task == task:
+                    # The scripts did not all finish. Stop them by turning them off.
+                    # Do not cancel them now. Stopping them should cause them to finish
+                    # gracefully.
+                    if service in (SERVICE_TOGGLE, SERVICE_TURN_ON):
+                        service_data = {
+                            key: value
+                            for key, value in service_data.items()
+                            if key in cv.ENTITY_SERVICE_FIELDS
+                        }
+                    else:
+                        service_data = {ATTR_ENTITY_ID: f"script.{service}"}
+                    await self._hass.services.async_call(
+                        "script", SERVICE_TURN_OFF, service_data, blocking=True
+                    )
+                else:
+                    pending_task.cancel()
+            # Propagate any exceptions that might have happened.
+            for done_task in done:
+                done_task.result()
+        else:
+            # No call limit, so just wait for it to finish.
+            await coro
 
-    async def async_run(self) -> None:
-        """Run script."""
-        self._hass.async_create_task(self._async_run(False))
 
+class _QueuedScriptRun(_ScriptRun):
+    """Manage queued Script sequence run."""
 
-class _BlockingScriptRun(_ScriptRun):
-    """Manage blocking Script sequence run."""
+    async def _async_run(self):
+        # pylint: disable=protected-access
+        _, pending = await asyncio.wait(
+            {self._stop.wait(), self._script._queue_lck.acquire()},
+            return_when=asyncio.FIRST_COMPLETED,
+        )
+        for pending_task in pending:
+            pending_task.cancel()
+        if self._stop.is_set():
+            self._finish()
+        else:
+            await super()._async_run()
 
-    async def async_run(self) -> None:
-        """Run script."""
+    def _finish(self):
+        # pylint: disable=protected-access
+        self._script._queue_len -= 1
         try:
-            await asyncio.shield(self._async_run())
-        except asyncio.CancelledError:
-            await self.async_stop()
-            raise
+            self._script._queue_lck.release()
```

Comment:
Isn't this dangerous as you could end up releasing for another run?

Reply:
The lock is shared by all runs of the _same script_. So if this run acquired the lock, it's safe (and the right thing to do) for it to release it, which will allow the next run to get it.

If, however, the script is stopped (i.e., the `self._stop.wait()` task finishes first), then _all_ runs of the script will also stop, too, so they they won't run anyway. I needed the try block to handle the case where the stop finished first.

So, bottom line, no, I believe this is correct.

Reply:
Actually, now that I think about it some more, the stop flag is unique to the run, not the script. Although in most cases they will all get set because the script will stop all runs at the same time, I suppose it's possible for a run to get cancelled by itself, and in this case it will not have acquired the lock (if it's still in this piece of code.) So, yeah, you're probably right. I'll fix it. Thanks!